Add Indexes via Schema

dplewis opened the pull request:
Pending: https://github.com/parse-community/parse-server/pull/4240

montymxb commented:
Cool! Same deal here as with the aggregate queries. Once the server side functionality is in I'll go over this.

dplewis commented:
@montymxb I added documentation and fixed the tests. The test weren't working locally at first until I updated my version of node from 6.10 to 9.2. Can you have a look into travis?

montymxb commented:
Sure thing @dplewis , I'll see what's going on here...

montymxb commented:
Took a look locally, seems to be just an outdated version of parse-server in cache. Dumped it and rerunning now. I did see one failure which looks to be consistent and is beyond this PR, probably due to a change in the server. We'll see if it reproduces when these run through.

dplewis commented:
Are you referring to the `testBadSchemaDelete` or `testCheckBadServer` tests?

montymxb commented:
Nope, schema tests look good actually. This is what I saw locally.
```
1) Parse\Test\ParseFileTest::testParseFileTypes
Failed asserting that two strings are equal.
--- Expected
+++ Actual
@@ @@
-'application/octet-stream'
+'null'
```
Curious if it's just a local issue or if it will repeat up here. If so I'll need to take a further look.

dplewis commented:
@montymxb I got that as well

montymxb commented:
Looks like there's still further issues starting up the test server. I'll have to look at this tomorrow when I have some time.

montymxb commented:
@dplewis master is good now, you can merge against it when you have a moment.

dplewis commented:
@montymxb thanks for fixing the tests

montymxb commented on src/Parse/ParseSchema.php:
```diff
@@ -348,6 +363,31 @@ public function addField($fieldName = null, $fieldType = 'String')
+      /**
+     * Adding an Index to Create / Update a Schema.
+     *
+     * @param string $indexName Name of the index will created on Parse
```
Bit of a typo
> Name of the index **that** will **be** created on Parse

montymxb commented on src/Parse/ParseSchema.php:
```diff
@@ -348,6 +363,31 @@ public function addField($fieldName = null, $fieldType = 'String')
+     *
+     * @return ParseSchema indexes return self to create index on Parse
+     */
+    public function addIndex($indexName = null, $index = null)
```
Not sure why default values of `null` are needed for both of the args. I think those defaults should be removed considering that a null value would cause this to throw an exception.

montymxb commented on src/Parse/ParseSchema.php:
```diff
@@ -614,6 +654,18 @@ public function deleteField($fieldName = null)
     }
 
+    /**
+     * Deleting a Index to Update on a Schema.
```
Deleting **an**  Index

montymxb commented on src/Parse/ParseSchema.php:
```diff
@@ -614,6 +654,18 @@ public function deleteField($fieldName = null)
+    /**
+     * Deleting a Index to Update on a Schema.
+     *
+     * @param string $indexName Name of the index will be deleted
```
**that** will be

montymxb commented on tests/Parse/ParseSchemaTest.php:
```diff
@@ -164,6 +164,23 @@ public function testUpdateSchemaStream()
+
+        $schema->addString('name');
+        $schema->update();
+        $schema->update();
```
Double update?

dplewis replied:
There was a bug where calling multiple updates would create multiple fields / indexes. I don't think this was intended functionality. Thus the `testUpdateMultipleSchemaStream`

montymxb replied:
Ahh, this is fine then.

montymxb commented on tests/Parse/ParseSchemaTest.php:
```diff
@@ -449,8 +466,108 @@ public function testBadSchemaDelete()
         $schema->delete();
     }
+
+    public function testCreateIndexSchemaStream()
```
I would consolidate these pair tests into singular tests. You don't need to manually test each as part of the CI selectively runs the entire test suite against the stream client in addition to the curl client. The other reason being if we ever add additional clients in the future we would have to add additional tests if we did it this way, versus just adding another CI run selectively using the alternate client instead. 

There are _some_ tests in here that still do this and it was probably me who added them. Those were early tests before I had setup the independent curl/stream CI runs. Honestly they can be selectively phased out at now. But to sum it up we shouldn't add additional stream differing tests (unless there's an extenuating circumstance) as it's redundant now.

montymxb replied:
Just to detail where it is you can see the stream client testing in the matrix for [**5.4** and **HHVM**](https://github.com/parse-community/parse-php-sdk/blob/master/.travis.yml#L17...L22) and later you can see how it is actually factored into the tests [here](https://github.com/parse-community/parse-php-sdk/blob/master/.travis.yml#L40...L41).

montymxb commented on tests/Parse/ParseSchemaTest.php:
```diff
@@ -449,8 +466,108 @@ public function testBadSchemaDelete()
+
+    public function testDeleteIndexSchema()
+    {
+        ParseClient::setHttpClient(new ParseStreamHttpClient());
```
Since this one is just on it's own you can just clear out the manual client.

montymxb commented on README.md:
````diff
@@ -679,6 +680,19 @@ A schema can be removed via `delete`, but it must be empty first.
+#### Index
+Indexes support efficient execution of queries from the database. MasterKey is required.
+```php
+// To add an index, the field must exist before you create an index
````
Could you add a few lines showing how to check existing indexes?

montymxb commented on tests/Parse/ParseSchemaTest.php:
```diff
@@ -411,7 +429,7 @@ public function testBadSchemaGet()
     public function testBadSchemaSave()
     {
-        $this->setExpectedException('\Parse\ParseException');
+        $this->setExpectedException('\Exception');
```
One last thing. Is there a particular reason you changed the expected exception here?

dplewis replied:
https://github.com/dplewis/parse-php-sdk/blob/schema-index/src/Parse/ParseSchema.php#L277

It throws exception instead of parse exception. The test were failing locally for some reason.

On a side note testBadSchemaDelete doesn't pass locally for me. If I remove the spaces from `$badClassName = "<Bad~ Class~ Name>` it passes.

montymxb replied:
Ahh that's something I missed. That change is good then. And quick question, are you running these tests on a windows machine by chance?

dplewis replied:
MacOS 10.12.6 Sierra

montymxb replied:
Hmm, good on mine and we're on the same setup. There's certainly a reason for it but I wonder what 🤔 . Anyways this is all good then.

dplewis replied:
Thanks for waiting. Now onto 1.4.0 👍

montymxb commented on tests/Parse/ParseSchemaTest.php:
```diff
@@ -427,7 +445,7 @@ public function testBadSchemaSave()
     public function testBadSchemaUpdate()
     {
-        $this->setExpectedException('\Parse\ParseException');
+        $this->setExpectedException('\Exception');
```
same question for this here too.

dplewis replied:
Same above

montymxb commented:
Sweet we're ready for **1.4.0**. I'll craft a release later tonight.
